pkg/util/aad/aad.go

```diff
@@ -39,15 +40,21 @@ func GetToken(ctx context.Context, log *logrus.Entry, clientID, clientSecret, te
 	// during credentials rotation this can take time to propagate
 	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
 	defer cancel()
-
 	// NOTE: Do not override err with the error returned by
 	// wait.PollImmediateUntil. Doing this will not propagate the latest error
 	// to the user in case when wait exceeds the timeout
 	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
 		var done bool
 		done, err = authorizer.RefreshWithContext(ctx, log)
 		if err != nil {
-			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
+			deadLine, _ := timeoutCtx.Deadline()
+			if azureerrors.IsInvalidSecretError(err) && !time.Now().After(deadLine.Add(-30*time.Second)) {
+				// Ignore the Invalid Secret errors until 30 seconds before the content timeout,
+				// because then we can hopefully assume it isn't caused by an AAD propagation delay
+				err = nil
+			} else {
+				err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
+			}
```

Comment:
The ticket says:

> I want receive meaningful error when I provide invalid clientid/clientsecret for CSP instead of Internal Server error

I'm not sure that I understand:

1. Why we were getting internal server error in the first place? Looks like we were returning `CloudError` which should be propogated to the customers.
2. How this change helps to solve the issue?

Reply:
@m1kola : 
1. we were receiving internal server error because the function RefreshWithContext (https://github.com/Azure/ARO-RP/blob/master/pkg/util/refreshable/refreshable.go) returned nil for 3 types of errors. So it will return sp, nil instead of errors in the parent function GetToken(https://github.com/Azure/ARO-RP/blob/master/pkg/util/aad/aad.go) which leads to a generic error message with wrong sp (servicePrincipalToken)

2. when we return error from RefreshWithContext for secret invalid error also, it will skip the wait and return error immediately, so inorder to add a propagation delay; Ignored the Invalid Secret errors until 30 seconds before the content timeout,  so that we can assume it isn't caused by an AAD propagation delay. This was suggested by @darthhexx , please correct me if i m wrong in the explanation

Reply:
> we were receiving internal server error because the function RefreshWithContext (https://github.com/Azure/ARO-RP/blob/master/pkg/util/refreshable/refreshable.go) returned nil for 3 types of errors. So it will return sp, nil instead of errors in the parent function GetToken(https://github.com/Azure/ARO-RP/blob/master/pkg/util/aad/aad.go) which leads to a generic error message with wrong sp (servicePrincipalToken)

Thanks for the explanation. I think I now have a better understanding, but the picture is not complete for me.

I'll repeat what you are saying just to make sure that we are on the same page:

1. [`GetToken`](https://github.com/Azure/ARO-RP/blob/93679fb7134afe26cd5daac527188df499551670/pkg/util/aad/aad.go#L23) calls `authorizer.RefreshWithContext(ctx, log)` inside `wait.PollImmediateUntil(...)`
2. `authorizer.RefreshWithContext(ctx, log)` in turn returns `done = false` and `err = nil`. This means that we hit `return` in this condition (it will be `return false, nil`):
    https://github.com/Azure/ARO-RP/blob/93679fb7134afe26cd5daac527188df499551670/pkg/util/aad/aad.go#L52-L54
3. The above return means that we try over and over again inside `wait.PollImmediateUntil` until we timeout. On the timeout, we hit this return:
    https://github.com/Azure/ARO-RP/blob/93679fb7134afe26cd5daac527188df499551670/pkg/util/aad/aad.go#L81
    Since there is no `err` (we deliberetely do not handle an error from `wait.PollImmediateUntil` for propogation reasons).

I think you got the root cause correct, but I'm curious - what happens next? if from `GetToken` we return `err = nil`, then what exactly fails and leads to 500/internal server error?

Reply:
Will something like this be better?

* Keep `RefreshWithContext` as is
* In `GetToken` (do we need to do the same enywhere else?) try to refresh with `wait.PollImmediateUntil`.
* Once we exit `wait.PollImmediateUntil` and if there are no errors - try to refresh one last time. In this case - if it errors - then it is for real and not due to propogation delay.

What do you think?

Reply:
@m1kola : suggested changes are made, thank you.